src/tribler/core/utilities/search_utils.py

```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+
+    Now some examples of how the function can be used. To use the function, you can call it one time for each word
+    from the query and see:
+    - how many query words are missed in the title;
+    - how many excess or out-of-place title words are found before each query word;
+    - and how many title words are not mentioned in the query.
+
+    Example 1, query "A B C", title "A B C":
+    find_term("A", deque(["A", "B", "C"])) -> (found=True, skipped=0, rest=deque(["B", "C"]))
+    find_term("B", deque(["B", "C"])) -> (found=True, skipped=0, rest=deque(["C"]))
+    find_term("C", deque(["C"])) -> (found=True, skipped=0, rest=deque([]))
+    Conclusion: exact match.
+
+    Example 2, query "A B C", title "A B C D":
+    find_term("A", deque(["A", "B", "C", "D"])) -> (found=True, skipped=0, rest=deque(["B", "C", "D"]))
+    find_term("B", deque(["B", "C", "D"])) -> (found=True, skipped=0, rest=deque(["C", "D"]))
+    find_term("C", deque(["C", "D"])) -> (found=True, skipped=0, rest=deque(["D"]))
+    Conclusion: minor penalty for one excess word in the title that is not in the query.
+
+    Example 3, query "A B C", title "X Y A B C":
+    find_term("A", deque(["X", "Y", "A", "B", "C"])) -> (found=True, skipped=2, rest=deque(["B", "C", "X", "Y"]))
+    find_term("B", deque(["B", "C", "X", "Y"])) -> (found=True, skipped=0, rest=deque(["C", "X", "Y"]))
+    find_term("C", deque(["C", "X", "Y"])) -> (found=True, skipped=0, rest=deque(["X", "Y"]))
+    Conclusion: major penalty for skipping two words at the beginning of the title plus a minor penalty for two
+    excess words in the title that are not in the query.
+
+    Example 4, query "A B C", title "A B X Y C":
+    find_term("A", deque(["A", "B", "X", "Y", "C"])) -> (found=True, skipped=0, rest=deque(["B", "X", "Y", "C"]))
+    find_term("B", deque(["B", "X", "Y", "C"])) -> (found=True, skipped=0, rest=deque(["X", "Y", "C"]))
+    find_term("C", deque(["X", "Y", "C"])) -> (found=True, skipped=2, rest=deque(["X", "Y"]))
+    Conclusion: average penalty for skipping two words in the middle of the title plus a minor penalty for two
+    excess words in the title that are not in the query.
+
+    Example 5, query "A B C", title "A C B":
+    find_term("A", deque(["A", "C", "B"])) -> (found=True, skipped=0, rest=deque(["C", "B"]))
+    find_term("B", deque(["C", "B"])) -> (found=True, skipped=1, rest=deque(["C"]))
+    find_term("C", deque(["C"])) -> (found=True, skipped=0, rest=deque(["C"]))
+    Conclusion: average penalty for skipping one word in the middle of the title.
+
+    Example 6, query "A B C", title "A C X":
+    find_term("A", deque(["A", "C", "X"])) -> (found=True, skipped=0, rest=deque(["C", "X"]))
+    find_term("B", deque(["C", "X"])) -> (found=False, skipped=0, rest=deque(["C", "X"]))
+    find_term("C", deque(["C", "X"])) -> (found=True, skipped=0, rest=deque(["X"]))
+    Conclusion: huge penalty for missing one query word plus a minor penalty for one excess title word.
```

Comment:
Thank you for the extensive explanation! Now it is clear what the function does. 🙏 

However, I would suggest you change the design of the function in the following way:
instead of modifying the argument, please return a triple `(found, skipped, rest)`. 
Then the parameters become immutable and the behavior of the function becomes more straightforward.

The current time complexity of the function is `O(n)+O(k)`

One of the options is to rewrite the function to 

```python
def find_term(term: str, title: List[str]) -> Tuple[bool, int, List]:
    accum = []

    for i, word in enumerate(title):
        if word == term:
            rest = title[i + 1::]
            rest.extend(accum)
            return True, i, rest
        else:
            accum.append(word)

    return False, 0, accum
```
It has the same time complexity: `O(n)+O(k)` but a better design.

The tests:

```python
def test_find_term():
    assert find_term("A", ["A", "B", "C"]) == (True, 0, ["B", "C"])
    assert find_term("B", ["B", "C"]) == (True, 0, ["C"])
    assert find_term("C", ["C"]) == (True, 0, [])

    assert find_term("A", ["X", "Y", "A", "B", "C"]) == (True, 2, ["B", "C", "X", "Y"])
    assert find_term("B", ["B", "C", "X", "Y"]) == (True, 0, ["C", "X", "Y"])
    assert find_term("C", ["C", "X", "Y"]) == (True, 0, ["X", "Y"])

    assert find_term("A", ["A", "C", "X"]) == (True, 0, ["C", "X"])
    assert find_term("B", ["C", "X"]) == (False, 0, ["C", "X"])
    assert find_term("C", ["C", "X"]) == (True, 0, ["X"])
```

Ref: https://wiki.python.org/moin/TimeComplexity

Reply:
This function is executed directly inside the SQL query and is potentially called for millions of torrents. In this case, the performance of the function directly affects the performance of the SQL query because the SQLite code itself works faster than the Python code executed inside the SQL code.

I did a quick benchmark of the function I use in PR and the function that you are suggesting:

```python
import timeit
t1 = timeit.Timer(
    "calculate_rank(['a', 'b', 'c', 'd', 'e'], ['x', 'a', 'b', 'y', 'z', 'c', 'd', 'e', 'f', 'g'])",
    "from tribler.core.utilities.search_utils import calculate_rank")
t2 = timeit.Timer(
    "calculate_rank2(['a', 'b', 'c', 'd', 'e'], ['x', 'a', 'b', 'y', 'z', 'c', 'd', 'e', 'f', 'g'])",
    "from tribler.core.utilities.search_utils import calculate_rank2")
t1.repeat()
t2.repeat()
```
The results are
```python
[2.4469536999999946, 2.4567987000000358, 2.4902578999999605, 2.5008845999999494, 2.5010604000000285]
```
for the function in the PR and
```python
[3.1153759999999693, 3.1840986999999927, 3.18589350000002, 3.1968425000000025, 3.2020163000000252]
```
for the function that you are suggesting.

So, the function that does not modify the title in place works 28% slower.

It is a pretty big difference and can lead to a 10% performance difference for the entire query. Tribler full-text searches are already slow; making them even slower just to simplify the function slightly is unnecessary. I think that the function in the PR is not so complex, and the user experience is more important here.

Reply:
@kozlovsky could you please give the full source of the test you performed?

Reply:
@drew2a sure, you can take it from here: https://pastebin.com/DTK6beT7

It shows the following results to me:
```python
For the test, time one million invocations of function, repeat 5 times
Original function speed: [2.6123462, 2.6423102, 2.6609092, 2.6560733, 2.6591641]
Suggested function speed: [3.365587399999999, 3.3657550999999977, 3.3585778000000026, 3.3598278000000015, 3.353927500000001]
The slowdown is 28.387558279986052 percent
```

Reply:
Thank you. 
Confirmed that your function is faster, so let's keep it.

But then we should rename it to something that reflects the fact that the `title` changes significantly during the function call.

Something `rotate_title_by_word`? I don't like this name, but maybe you will come up with something better.

Reply:
I agree that the name should reflect the fact that the function mutates the title.
I renamed it from `find_word` to `find_word_and_rotate_title`

Reply:
🙏